/*
 * @begin:license
 *
 * Copyright (c) 2015-2018, Benjamin Niemann <pink@odahoda.de>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * @end:license
 */

syntax = "proto2";

package noisicaa.pb;

message NotSupportedReasons {
  repeated string unsupported_lv2_feature = 1;
  repeated string unsupported_lv2_ui_type = 2;
}

message ProcessorDescription {
  enum Type {
    NULLPROC = 1;
    CSOUND = 2;
    CUSTOM_CSOUND = 3;
    FLUIDSYNTH = 4;
    PLUGIN = 5;
    SAMPLE_PLAYER = 6;
    SOUND_FILE = 7;
    TRACK_MIXER = 8;
    PIANOROLL = 9;
    CV_GENERATOR = 10;
    SAMPLE_SCRIPT = 11;
  }
  required Type type = 1;
}

message LV2Feature {
  required string uri = 1;
  required bool required = 2;
}

message LV2Description {
  required string uri = 1;

  message UI {
    required bool supported = 1;
    optional NotSupportedReasons not_supported_reasons = 2;

    required string uri = 3;
    optional string type_uri = 4;
    optional string bundle_path = 5;
    optional string binary_path = 6;
    repeated LV2Feature features = 7;
  }

  repeated UI uis = 2;
  optional string ui_uri = 3;

  repeated LV2Feature features = 4;
}

message LadspaDescription {
  required string library_path = 1;
  required string label = 2;
}

message CSoundDescription {
  required string orchestra = 1;
  required string score = 2;
}

message FluidSynthDescription {
  required string soundfont_path = 1;
  required uint32 bank = 2;
  required uint32 preset = 3;
}

message SamplePlayerDescription {
  required string sample_path = 1;
}

message SoundFileDescription {
  required string sound_file_path = 1;
}

message PluginDescription {
  enum Type {
    LV2 = 1;
    LADSPA = 2;
  }
  required Type type = 1;
}

message FloatValueDescription {
  optional float default = 1;
  optional float min = 2;
  optional float max = 3;
}

message PortDescription {
  enum Direction {
    INPUT = 1;
    OUTPUT = 2;
  }
  required Direction direction = 1;

  enum Type {
    UNSUPPORTED = -1;
    AUDIO = 1;
    ARATE_CONTROL = 2;
    KRATE_CONTROL = 3;
    EVENTS = 4;
  }
  required Type type = 2;

  required string name = 3;
  optional string display_name = 4;

  optional FloatValueDescription float_value = 5;
  optional string csound_instr = 6;
  optional string drywet_port = 7;
  optional float drywet_default = 8;
  optional string bypass_port = 9;
}

message NodeDescription {
  optional bool supported = 1 [default = true];
  optional NotSupportedReasons not_supported_reasons = 2;

  enum Type {
    OTHER = 1;
    PROCESSOR = 2;
    PLUGIN = 3;
    REALM_SINK = 4;
    CHILD_REALM = 5;
    EVENT_SOURCE = 6;
  }
  required Type type = 3;

  optional string display_name = 4;

  optional bool has_ui = 5 [default = false];

  repeated PortDescription ports = 6;

  optional ProcessorDescription processor = 100;

  // Processor types
  optional CSoundDescription csound = 200;
  optional FluidSynthDescription fluidsynth = 201;
  optional SamplePlayerDescription sample_player = 202;
  optional SoundFileDescription sound_file = 203;
  optional PluginDescription plugin = 204;

  // Plugin types
  optional LV2Description lv2 = 300;
  optional LadspaDescription ladspa = 301;
}
